src/main/java/org/mvel2/util/ParseTools.java

```diff
@@ -2210,4 +2232,88 @@ public static Class findInnerClass( String className, ClassLoader classLoader, C
     }
     throw cnfe;
   }
+
+  public static Class findInnerClassCached(String className, ClassLoader classLoader, ClassNotFoundException cnfe) throws ClassNotFoundException {
+    Class clazz;
+    for (int lastDotPos = className.lastIndexOf('.'); lastDotPos > 0; lastDotPos = className.lastIndexOf('.')) {
+      className = className.substring(0, lastDotPos) + "$" + className.substring(lastDotPos + 1);
+      if (CLASS_NOT_FOUND_MAP.containsKey(className)) {
+        continue;
+      }
+      clazz = CLASS_MAP.get(className);
+      if (clazz == null) {
+        try {
+          clazz = classLoader.loadClass(className);
+          CLASS_MAP.put(className, clazz);
+          return clazz;
+        } catch (ClassNotFoundException e) {
+          CLASS_NOT_FOUND_MAP.put(className, Boolean.TRUE);
+        }
+      }
+    }
+    throw cnfe;
+  }
+
+  public static Class forNameCached(String className, ClassLoader classLoader) throws ClassNotFoundException {
+    if(CLASS_NOT_FOUND_MAP.containsKey(className)){
+      throw new ClassNotFoundException(className);
+    }
+    Class<?> clazz = CLASS_MAP.get(className);
```

Comment:
The class loader is not part of the key. A class might end up in the CLASS_NOT_FOUND set if it's first loaded from a class loader, then it might be available when called from another class loader but this will not catch it. This could happen for example when running as a webapp in Tomcat (some classes are provided by the container, others by the webapp)